rflx/model.py

```diff
@@ -546,68 +541,79 @@ def __check_length_expression(self, link: Link) -> None:
 
     def __prove_conflicting_conditions(self) -> None:
         for f in (INITIAL, *self.__fields):
-            conflict = LessEqual(
-                Add(
-                    *[
-                        If([(self.__with_constraints(c.condition), Number(1))], Number(0))
-                        for c in self.outgoing(f)
-                    ]
-                ),
-                Number(1),
-            )
-            result = conflict.forall()
-            if result != ProofResult.sat:
-                message = str(conflict).replace("\n", "")
-                raise ModelError(
-                    f'conflicting conditions for field "{f.name}"'
-                    f' in "{self.identifier}" ({result}: {message})'
-                )
+            for i1, c1 in enumerate(self.outgoing(f)):
+                for i2, c2 in enumerate(self.outgoing(f)):
+                    if i1 != i2:
+                        conflict = And(c1.condition, c2.condition)
+                        proof = conflict.check(self.__type_constraints(conflict))
+                        if proof.result == ProofResult.sat:
+                            raise ModelError(
+                                f'conflicting conditions {i1} and {i2} for field "{f.name}"'
+                                f' in "{self.identifier}"'
+                            )
 
     def __prove_reachability(self) -> None:
+        def has_final(field: Field) -> bool:
+            if field == FINAL:
+                return True
+            for o in self.outgoing(field):
+                if has_final(o.target):
+                    return True
+            return False
+
+        for f in (INITIAL, *self.__fields):
+            if not has_final(f):
+                raise ModelError(f'no path to FINAL for field "{f.name}"')
+
         for f in (*self.__fields, FINAL):
-            reachability = Or(
-                *[
-                    And(*[self.__with_constraints(l.condition) for l in path])
-                    for path in self.__paths[f]
-                ]
-            )
-            result = reachability.exists()
-            if result != ProofResult.sat:
-                message = str(reachability).replace("\n", "")
+            errors = []
+            found = False
+            for path in self.__paths[f]:
+                facts = [fact for link in path for fact in self.__link_expression(link)]
+                if f != FINAL:
+                    facts.extend(
+                        self.__expression_list(
+                            Or(*[o.condition for o in self.outgoing(f)]).simplified()
+                        )
+                    )
+                proof = TRUE.check(facts)
+                if proof.result == ProofResult.sat:
+                    found = True
+                else:
+                    path_message = " -> ".join([l.target.name for l in path])
+                    errors.append(f"[{path_message}]:\n   {proof.error}")
```

Comment:
Could this be
```Python
found = proof.result == ProofResult.sat
if not found:
    path_message = " -> ".join([l.target.name for l in path])
    errors.append(f"[{path_message}]:\n   {proof.error}")
```
?
Otherwise it looks as if the inner loop should break when when `proof.result == ProofResult.sat` because the outer one only checks if `found` is false.

Reply:
I realize that the code is confusing. The idea is that at least one path to a field has its conditions satisfied (i.e. is reachable). The `found` variable indicated that one such satisfied condition exists, but the naming is unfortunate in that it hints something like search algorithm.

In 636faa44f152145acc06f2f9378316d030446e99 I simplified the algorithm to exit the loop as soon as a satisfied path is found. This should be clearer and also more efficient as unnecessary proofs are avoided.